### PR TITLE
Fix missing steps bug in is-fairworkflows decorator

### LINKED ISSUE #175: Missing steps bug in is-fairworkflow decorator
The steps are completely missing from workflows constructed using the `is_fairworkflow` decorator.

### PR DESCRIPTION
**Pull request by svenvanderburg**
See the addition to the test, the FairWorkflow object resulting from specifying the workflow with `is_fairworkflow` decorator did not have any steps. This is because apparently the PromisedObject created out of the scheduled workflow does not hold information about the individual steps. See here my horrible solution. Since this noodles execution is super horrible code anyway I don't think we should spend a lot of time in making this nicer at this point.

Fixes #175

### CODE REVIEW

File: fairworkflows/fairworkflow.py [hunk 1]
```diff
@@ -86,24 +86,29 @@ def from_function(cls, func: Callable):
...
                    derived_from=derived_from, language=LINGSYS_PYTHON)
-        self.noodles_promise = noodles_promise
+        self.workflow_level_promise = workflow_level_promise
 
-        workflow = noodles.get_workflow(self.noodles_promise)
+        workflow = noodles.get_workflow(step_level_promise)
 
-        # Exclude the root, because that is the workflow definition itself and not a step
-        steps_dict = {i: n.foo._fairstep for i, n in workflow.nodes.items() if i != workflow.root}
+        steps_dict = {i: n.foo._fairstep for i, n in workflow.nodes.items()}
```
Comment (raar1) [inline fairworkflows/fairworkflow.py:106]: I confess I've sort of forgotten what the reasoning was here - so we're now including the workflow 'root' step?

File: tests/test_fairworkflow.py [hunk 2]
```diff
@@ -347,6 +347,7 @@ def my_workflow(in1, in2, in3):
         fw = FairWorkflow.from_function(my_workflow)
 
         assert isinstance(fw, FairWorkflow)
+        assert len(fw._steps) == 4
```
Comment (raar1) [inline tests/test_fairworkflow.py:350]: This is nice, although I suppose this wouldn't work if we 'reused' a step? I think that's already its own issue though.

File: fairworkflows/fairworkflow.py [hunk 3]
```diff
@@ -86,24 +86,29 @@ def from_function(cls, func: Callable):
                              'use is_fairworkflow decorator to mark it.')
 
     @classmethod
-    def from_noodles_promise(cls, noodles_promise, description: str = None, label: str = None,
-                 is_pplan_plan: bool = True, derived_from=None):
+    def from_noodles_promise(cls, workflow_level_promise, step_level_promise, description: str = None, label: str =
```
Comment (raar1) [inline fairworkflows/fairworkflow.py:89]: Perhaps we should add type hints here to ensure noodles promises are being passed?

**Review by raar1 — APPROVED** [review]
Nice - this is also clearer than what was there before. I've left one or two minor comments

File: fairworkflows/fairworkflow.py (hunk 1 above)
Comment (svenvanderburg) [reply fairworkflows/fairworkflow.py:106]: We're now using the `step_level_promise` which does not have the reference to the function defining the workflow itself. The root node is thus one of the steps and we want to include it.

File: tests/test_fairworkflow.py (hunk 2 above)
Comment (svenvanderburg) [reply tests/test_fairworkflow.py:350]: If this behavior changes we shall update the test.

File: fairworkflows/fairworkflow.py (hunk 3 above)
Comment (svenvanderburg) [reply fairworkflows/fairworkflow.py:89]: Added :)

**Pull request merged**